cfgmgr/coppmgr.cpp

```diff
@@ -0,0 +1,859 @@
+                    if (m_coppTrapIdTrapGroupMap.find(it.first) != m_coppTrapIdTrapGroupMap.end())
+                    {
+                        string trap_group = m_coppTrapIdTrapGroupMap[it.first];
+
+                        /* Delete the restricted copp group when it goes to pending state
+                         * on feature disable
+                         */
+                        if ((m_coppGroupRestrictedMap.find(trap_group) 
+                             != m_coppGroupRestrictedMap.end()) &&
+                             (checkIfTrapGroupFeaturePending(trap_group)))
+                        {
+                            m_appCoppTable.del(trap_group);
+                            delCoppGroupStateOk(trap_group);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+bool CoppMgr::isTrapDisabled(string trap_id)
+{
+    auto trap_idx = m_coppTrapDisabledMap.find(trap_id);
+    if (trap_idx != m_coppTrapDisabledMap.end())
+    {
+        return m_coppTrapDisabledMap[trap_id];
+    }
+    return false;
+}
+
+/* The genetlink  fields are restricted and needs to be installed only on
+ * feature(sflow) enable
+ */
+bool CoppMgr::coppGroupHasRestrictedFields (vector<FieldValueTuple> &fvs)
+{
+    for (auto i: fvs)
+    {
+        if ((fvField(i) == COPP_GROUP_GENETLINK_NAME_FIELD) ||
+            (fvField(i) == COPP_GROUP_GENETLINK_MCGRP_NAME_FIELD))
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+CoppMgr::CoppMgr(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, const vector<string> &tableNames) :
+        Orch(cfgDb, tableNames),
+        m_cfgCoppTrapTable(cfgDb, CFG_COPP_TRAP_TABLE_NAME),
+        m_cfgCoppGroupTable(cfgDb, CFG_COPP_GROUP_TABLE_NAME),
+        m_cfgFeatureTable(cfgDb, CFG_FEATURE_TABLE_NAME),
+        m_appCoppTable(appDb, APP_COPP_TABLE_NAME),
+        m_stateCoppTrapTable(stateDb, STATE_COPP_TRAP_TABLE_NAME),
+        m_stateCoppGroupTable(stateDb, STATE_COPP_GROUP_TABLE_NAME),
+        m_coppTable(appDb, APP_COPP_TABLE_NAME)
+{
+    SWSS_LOG_ENTER();
+    parseInitFile();
+    std::vector<string> group_keys;
+    std::vector<string> trap_keys;
+    std::vector<string> feature_keys;
+    std::vector<string> app_keys;
+
+    std::vector<string> group_cfg_keys;
+    std::vector<string> trap_cfg_keys;
+
+    CoppCfg group_cfg;
+    CoppCfg trap_cfg;
+
+    m_cfgCoppGroupTable.getKeys(group_cfg_keys);
+    m_cfgCoppTrapTable.getKeys(trap_cfg_keys);
+    m_cfgFeatureTable.getKeys(feature_keys);
+    m_coppTable.getKeys(app_keys);
+
+
+    for (auto it: trap_feature_map)
+    {
+        m_coppTrapDisabledMap[it.first] = true;
+    }
+    for (auto i: feature_keys)
+    {
+        std::vector<FieldValueTuple> feature_fvs;
+        m_cfgFeatureTable.get(i, feature_fvs);
+
+        bool status = false;
+        for (auto j: feature_fvs)
+        {
+            if (fvField(j) == "status")
```

Comment:
We can merge this kind of 2 conditions.  ( fvField(j) == "status" && fvValue(j) == "enabled" )

Reply:
Done